### ISSUE TITLE
Understanding the behavior of pconnect

### ISSUE THREAD

**Issue by drpoggi** [ibmdb/python-ibmdb#314]
Hello all. I'm seeing some interesting behavior from the `ibm_db.pconnect` method and I was wondering if you could help me understand if this is expected, or possibly a bug.

**Client environment:**
Ubuntu 14.04
Seen on Python 2.7 and 3.6

```
>>> ibm_db.__version__
'2.0.8'
```

**Database Server**
```
>>> info = ibm_db.server_info(conn)
>>> print(info.DBMS_NAME, info.DBMS_VER)
AS 07.02.0000
```

We use subsystems on the AS400 to control our jobs, and the UID we use to connect is routed to that user's subsystem. From what I can tell, a connection from the client is handled by an individual job on the server. If I kill the server job, I can see the connection on the client report as `CLOSE_WAIT`. If I then call `ibm_db.active(conn)` I get `False`, which appears correct. The issue occurs that a subsequent call to `pconnect` returns the same (closed) connection object, and when I try to use that connection object for something like `callproc` I get the error:

- `Exception: Describe Param Failed: [IBM][CLI Driver] SQL0900N  The application state is in error.  A database connection does not ex ist.  SQLSTATE=08003 SQLCODE=-900`

I would expect `pconnect` to detect the connection was closed and return a new connection object, however it seems to continue to return the closed connection object. Here are my steps to make that more clear:

```
>>> conn = ibm_db.pconnect('DATABASE=dbname;HOSTNAME=host;PORT=446;PROTOCOL=TCPIP;UID=user;PWD=password;', "", "", {3 017: ibm_db.SQL_TRUE})
>>> conn
<ibm_db.IBM_DBConnection object at 0x7f817d64bec0>
>>> ibm_db.active(conn)
True
>>> os.getpid()
2246
```

```
$ sudo lsof -p 2246 | grep dev.host.com:
python  2246 vagrant    3u  IPv4  81896      0t0    TCP 10.0.2.15:45728->dev.host.com:446 (ESTABLISHED)
```
(any usage of the `conn` here works)
- Kill Job associated with that connection on the AS400
```
$ sudo lsof -p 2246 | grep dev.host.com
python  2246 vagrant    3u  IPv4  81896      0t0    TCP 10.0.2.15:45728->dev.host.com:446 (CLOSE_WAIT)
```
```
>>> ibm_db.active(conn)
False
>>> ibm_db.close(conn)
True
```
`$ sudo lsof -p 2246 | grep dev.host.com` 
reports no open connections

`>>> del conn`
```
>>> conn = ibm_db.pconnect('DATABASE=dbname;HOSTNAME=host;PORT=446;PROTOCOL=TCPIP;UID=user;PWD=password;', "", "", {3017: ibm_db.SQL_TRUE})
>>> conn
<ibm_db.IBM_DBConnection object at 0x7f817d64bec0>
```
Note the connection object ID is the same as it was above
```
>>> res = ibm_db.callproc(conn, stored_procedure_name, params)
Exception: Describe Param Failed: [IBM][CLI Driver] SQL0900N  The application state is in error.  A database connection does not ex ist.  SQLSTATE=08003 SQLCODE=-900
```
Call again:
```
>>> res = ibm_db.callproc(conn, stored_procedure_name, params)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
Exception: [IBM][CLI Driver] CLI0106E  Connection is closed. SQLSTATE=08003 SQLCODE=-99999
```

At this point subsequent calls to `pconnect` will continue to return the same object, despite `ibm_db.active(conn)` returning `False`. I would expect `pconnect` to know the connection was closed and return a new, active connection. What is the expected behavior here? The fact that I can `del` the connection object and I still get back the same connection object ID makes me think this must be some kind of bug in the C layer?

Hopefully I've clearly articulated my problem, thanks for any help you can provide.

**Comment by SabaKauser** [comment]
Hello David,
I havn't looked into this yet. I shall get back to you after investigation.

**Comment by SabaKauser** [comment]
Wanted to know if you are able to reproduce this behavior with Db2 on Linux? We do not have iSeries machines to try this out.

**Comment by drpoggi** [comment]
I've dug into this a little more and believe I've found what was causing this bug. I don't think it's specific to the iSeries (although I don't have any other DB2 environment to test against), it seems like it would happen in any case where a connection object returned from `pconnect` had the connection closed out from under it. 

If you look at: https://github.com/ibmdb/python-ibmdb/blob/master/IBM_DB/ibm_db/ibm_db.c#L1189. Note it's setting `reused = 1` outside of the checks for `SQL_SUCCESS` and `conn_alive`, which seems wrong. The comment right above this line says "else will re-connect since connection is dead", but all of the connection logic following this line only happens if `reused=0`

I tested locally with the linked line removed and now pconnect works as I would expect it to (and how it should according to that comment.)
